### PR TITLE
Run migrations once per package 

### PR DESCRIPTION
**Pull request by camdencheek**
Previously, every time `NewDB()` was called, we would create a new
database and run the migration. We can significantly reduce the work
done by `NewDB()` by running the schema migration once, then creating new
test DBs using the first DB as a template. Creating a new DB from a
template takes ~500ms on my machine vs 6.5 seconds when running the
migrations. 

Note that this implementation creates a new template database per
package. Since go runs tests for each package as a separate process and
in parallel, it would be difficult to synchronise creating the template
DB across packages. However, I'd love suggestions to make this better.

This change will continue to improve performance as we migrate from
`dbtesting` to `dbtest`. And, since this makes creating DBs cheaper,
it encourages more parallelization. 

Before:
```
❯ go test ./internal/database
ok      github.com/sourcegraph/sourcegraph/internal/database    110.768s
```

After:
```
❯ go test ./internal/database
ok      github.com/sourcegraph/sourcegraph/internal/database    82.559s
```

### CODE REVIEW

File: internal/database/dbtest/dbtest.go [hunk 1]
```diff
@@ -96,6 +87,56 @@ func NewDB(t testing.TB, dsn string) *sql.DB {
...
+var templateOnce sync.Once
+
+// initTemplateDB creates a template database with a fully migrated schema for the
+// current package. New databases can then do a cheap copy of the migrated schema
+// rather than running the full migration every time.
+func initTemplateDB(t testing.TB, config *url.URL) {
+	templateOnce.Do(func() {
+		templateName := templateDBName()
+		db := dbConn(t, config)
+		dbExec(t, db, `DROP DATABASE IF EXISTS `+pq.QuoteIdentifier(templateName))
```
Comment (camdencheek) [inline internal/database/dbtest/dbtest.go:99]: Is there a good way to `CREATE DATABASE IF NOT EXISTS`? Right now, I'm dropping and recreating even though we could potentially cut out the 6 second delay for every test run by re-using a template db from a previous run.

File: internal/database/dbtest/dbtest.go [hunk 2]
```diff
@@ -96,6 +87,56 @@ func NewDB(t testing.TB, dsn string) *sql.DB {
...
+		} {
+			m, err := dbconn.NewMigrate(templateDB, database)
+			if err != nil {
+				t.Fatalf("failed to construct migrations: %s", err)
+			}
+			if err = dbconn.DoMigrate(m); err != nil {
+				t.Fatalf("failed to apply migrations: %s", err)
+			}
+		}
+		dbExec(t, db, killClientConnsQuery, templateName)
```
Comment (camdencheek) [inline internal/database/dbtest/dbtest.go:120]: Interestingly, `templateDB.Close()` did not seem to immediately end all connections to the database, so I had to kill remaining connections manually. Anyone know why that might be?

File: internal/database/dbtest/dbtest.go (hunk 1 above)
Comment (mrnugget) [reply internal/database/dbtest/dbtest.go:99]: Off the top of my head: you could try to shell out to `psql` and do a `psql -c "SELECT 1 FROM pg_database WHERE datname = "+templateName || psql create-database` (excuse the pseudocode syntax).

**Review by mrnugget — APPROVED** [review]
Nice! Bravo! 

Do  you know how this stacks up against `dbtesting.SetupGlobalTestDB` performance-wise? Because that's what we've been using in the batch changes packages. It was faster to setup a global test db once per package (and use it carefully) than to use `dbtest.NewDB` in every test.

https://github.com/sourcegraph/sourcegraph/blob/03a66dcf0ac85e6fb891c3d48cb53a8d6504324e/internal/database/dbtesting/dbtesting.go#L55-L62

**Review by ryanslade — APPROVED** [review]
Nice!

**Review by eseliger — APPROVED** [review]
Nice, that's a huge win :)

**Comment by camdencheek** [comment]
@mrnugget Anecdotally, converting [internal/database/search_contexts_test.go](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/search_contexts_test.go?subtree=true) to use `NewDB()` and `t.Parallel()` rather than `dbtesting.SetupGlobalTestDB()` cut runtime for those tests from 14s to 4s (9 tests total, each calling `dbtesting.SetupGlobalTestDB()`. Now, if you're only calling `dbtesting.SetupGlobalTestDB()` once _per package_, it is likely still faster than running `NewDB()` once _per test_, but YMMV, especially if you also parallelize tests that couldn't be before

**Comment by camdencheek** [comment]
In addition to this, I plan to spend a little time looking at maintaining a pool of migrated DBs that we can just truncate and use. Not sure if performance will actually be better than the 0.5s template clone, but it seems worth taking a look at if we can make creating a test db negligibly cheap

File: internal/database/dbtest/dbtest.go (hunk 2 above)
Comment (unknwon) [reply internal/database/dbtest/dbtest.go:120]: IIRC, the golang-migrate.Migrate holds the underlying database connection, so we need to do `m.Close()` for each `m` in the above loop.

**Review by unknwon — COMMENTED** [review]
Nice!

**Pull request merged**